f90wrap_helpers/classes.py

```diff
@@ -0,0 +1,84 @@
+import numpy as np
+
+from pyhmcode.hmx import HMcode2015 as HMcode2015
+from pyhmcode.hmx import HMcode2016 as HMcode2016
+from pyhmcode.hmx import HMcode2020 as HMcode2020
+from pyhmcode.hmx import HMcode2020_feedback as HMcode2020_feedback
+
+from pyhmcode.hmx import HMx2020_matter_pressure_w_temp_scaling as HMx2020_matter_pressure_w_temp_scaling
+from pyhmcode.hmx import HMx2020_matter_w_temp_scaling as HMx2020_matter_w_temp_scaling
+
+from pyhmcode.hmx import field_dmonly as field_dmonly
+from pyhmcode.hmx import field_matter as field_matter
+from pyhmcode.hmx import field_cdm as field_cdm
+from pyhmcode.hmx import field_gas as field_gas
+from pyhmcode.hmx import field_stars as field_stars
+from pyhmcode.hmx import field_electron_pressure as field_electron_pressure
+
+class Cosmology(pyhmcode.cosmology_functions.cosmology):
+    def __init__(self):
+        super().__init__()
+
+        # Assign reasonable default parameters (icosmo=1 == "boring cosmology")
+        pyhmcode.cosmology_functions.assign_cosmology(icosmo=1, cosm=self, verbose=False)
+
+        self.norm_method = pyhmcode.cosmology_functions.norm_none
+        self.itk = pyhmcode.cosmology_functions.iTk_external
+        self.has_power = True
+
+        self._has_linear_power_set = False
+
+    def set_linear_power_spectrum(self, k, z, pofk_lin):
+        if pofk_lin.shape != (len(z), len(k)):
+            raise ValueError(f"Shape mismatch between k, z, pofk_lin: {k.shape}, {z.shape}, {pofk_lin.shape}")
+        if z[0] >= z[-1]:
+            raise ValueError("Redshifts need to be in increasing order.")
+
+        pk_lin = np.asfortranarray(pofk_lin[::-1].T, dtype=np.float64)
+        self.a_lin = np.asfortranarray((1/(1+z))[::-1], dtype=np.float64)
+        self.k_lin = np.asfortranarray(k, dtype=np.float64)
+
+        pyhmcode.cosmology_functions.init_external_linear_power_tables(self, self.k_lin, self.a_lin, pk_lin)
+        pyhmcode.cosmology_functions.init_cosmology(self)
+
+        self._has_linear_power_set = True
+
+class Halomodel(pyhmcode.hmx.halomod):
+    def __new__(cls, mode, verbose=False):
+        # Need to use __new__ to intercept creation of the instance and use
+        # assign_halomod instead
+        instance = pyhmcode.hmx.assign_halomod(mode, verbose)
+        return instance
+
+
+
+def calculate_nonlinear_power_spectrum(cosmology, halomodel, fields=np.array([field_dmonly]),
+                                       return_halo_terms=False, verbose=False):
+    if not cosmology._has_linear_power_set:
+        raise RuntimeError("Cosmology has no linear power spectrum set.")
+
+    nk = len(cosmology.k_lin)
+    na = len(cosmology.a_lin)
+    nf = len(fields)
+
+    # Output arrays
+    pow_lin = np.zeros((nk, na), dtype=np.float64, order="F")
```

Comment:
Could we change `pow_lin` to `pow_li` just to make all the variables the same length to ease reading.

Reply:
It's `lin` everywhere else. I prefer the extra semantic meaning over the extra pretty aligment.